packages/foam-workspace-manager/src/utils/utils.ts

```diff
@@ -1,23 +1,103 @@
-// @todo convert this to use ast parsing
-
-// import unified from 'unified';
-// import markdown from 'remark-parse';
-// import wikiLinkPlugin from 'remark-wiki-link';
-// let processor = unified()
-//   .use(markdown, { gfm: true })
-//   .use(wikiLinkPlugin);
-
-import { findTopLevelHeading, findWikilinksInMarkdown } from './markdown-utils';
+import * as path from 'path';
+import unified from 'unified';
+import markdownParse from 'remark-parse';
+import wikiLinkPlugin from 'remark-wiki-link';
+import visit, { CONTINUE, EXIT } from 'unist-util-visit';
+import { Node, Parent } from 'unist';
+import { BubbleLink, Bubble, Foam } from '../core';
 
 // @ts-expect-error
 export function readWorkspaceFile(filename: string): string {
   throw new Error('Not implemented');
 }
 
+// pipeline cache
+let processor: unified.Processor | null = null;
+
+function parse(markdown: string): Node {
+  processor =
+    processor ||
+    unified()
+      .use(markdownParse, { gfm: true })
+      .use(wikiLinkPlugin);
+  return processor.parse(markdown);
+}
+
+export function createBubbleFromMarkdown(uri: string, markdown: string): Bubble {
+  const filename = path.basename(uri);
+  const id = path.parse(filename).name;
+  const tree = parse(markdown)
+  let title = id
+  visit(tree, node => {
+    if (node.type === 'heading' && node.depth === 1) {
+      title = ((node as Parent)!.children[0].value as string) || title;
+    }
+    return title === id ? CONTINUE : EXIT;
+  })
+  const links: BubbleLink[] = []
+  visit(tree, node => {
+    if (node.type === 'wikiLink') {
+      links.push({
+        from: id,
+        to: node.value as string,
+        text: node.value as string,
+      });
+    }
+  });
+  return new Bubble(id, title, links, uri, markdown)
+}
+
+interface MarkdownReference {
+  linkText: string
+  wikiLink: string
+  pageTitle: string
+}
+
+export function createMarkdownReferences(foam: Foam, bubbleId: string): MarkdownReference[] {
```

Comment:
I moved this here from the extension, this way it can also be used e.g. by the cli